File: leetcode/Python/0261-m-graph-valid-tree.py
"""
bfs
"""
import collections
class Solution_1:
    def validTree(self, n: int, edges: List[List[int]]) -> bool:
        if len(edges) != n - 1:
            return False
        
        graph = collections.defaultdict(list)
        for u, v in edges:
            graph[u].append(v)
            graph[v].append(u)
        
        visited = set()
        q = collections.deque([0])
        while q:
            curr = q.popleft()
            visited.add(curr)
            for node in graph[curr]:
                if node not in visited:
                    visited.add(node)
                    q.append(node)
        
        return len(visited) == n

"""
dfs
"""
class Solution_2:
    def validTree(self, n: int, edges: List[List[int]]) -> bool:
        graph = collections.defaultdict(list)
        for u, v in edges:
            graph[u].append(v)
            graph[v].append(u)
        visited = set()

        if not self.helper(0, -1, graph, visited):
            return False

        return len(visited) == n

    def helper(self, curr, parent, graph, visited):
        if curr in visited:
            return False
        visited.add(curr)
        for i in graph[curr]:
            if i != parent and not self.helper(i, curr, graph, visited):
                return False
        return True

"""
union find
"""
class Solution_3:
    def validTree(self, n: int, edges: List[List[int]]) -> bool:
        if not len(edges) == n - 1:
            return False
        
        parent = [-1] * n
        for u, v in edges:
            x = self.find_parent(parent, u)
            y = self.find_parent(parent, v)
            if x == y:
                return False
            else:
                self.union(parent, x, y)
        
        return True

    # find the subset of an element i
    def find_parent(self, parent, i):
        if parent[i] == -1:
            return i
        else:
            return self.find_parent(parent, parent[i])

    # do union of two subsets
    def union(self, parent, x, y):
        x_set = self.find_parent(parent, x)
        y_set = self.find_parent(parent, y)
        parent[x_set] = y_set

File: leetcode/C++/0261-m-graph-valid-tree.cpp
// DFS
class Solution {
private:
    bool hasCycle(unordered_map<int, vector<int>>& graph, int cur, int pre, vector<bool>& visited) {
        if (visited[cur]) return true;
        visited[cur] = true;
        for (int next : graph[cur]) {
            if (next != pre && hasCycle(graph, next, cur, visited))
                return true;
        }
        return false;
    }
        
public:
    bool validTree(int n, vector<vector<int>>& edges) {
        unordered_map<int, vector<int>> graph;
        vector<bool> visited(n, false);
        for (int i = 0; i < edges.size(); i++) {
            graph[edges[i][0]].push_back(edges[i][1]);
            graph[edges[i][1]].push_back(edges[i][0]);
        }
        
        if (hasCycle(graph, 0, -1, visited))
            return false;
        
        for (bool v : visited)
            if (!v) return false;
        
        return true;
    }
};

// Union Find
class Solution {
public:
    bool validTree(int n, vector<vector<int>>& edges) {
        if (edges.size() != n - 1)
            return false;
        
        vector<int> parent(n, -1);
        int x, y;
        for (int i = 0; i < n - 1; i++) {
            x = find_parent(parent, edges[i][0]);
            y = find_parent(parent, edges[i][1]);
            if (x == y)
                return false;
            else
                union_op(parent, x, y);
        }
        
        return true;
    }
    
    int find_parent(vector<int>& parent, int i) {
        if (parent[i] == -1)
            return i;
        return find_parent(parent, parent[i]);
    }
    
    void union_op(vector<int>& parent, int x, int y) {
        int x_set = find_parent(parent, x);
        int y_set = find_parent(parent, y);
        
        if (x_set != y_set)
            parent[x_set] = y_set;
    }
};